File: src-tauri/src/types/chunks.rs
/// Chunk types for session visualization.
/// Matches the TypeScript types in `src/main/types/chunks.ts`.

use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::domain::SessionMetrics;
use super::messages::{ParsedMessage, ToolCall, ToolResult};

// =============================================================================
// Process (Subagent Execution)
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Process {
    pub id: String,
    pub file_path: String,
    pub messages: Vec<ParsedMessage>,
    pub start_time: String,
    pub end_time: String,
    pub duration_ms: f64,
    pub metrics: SessionMetrics,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub subagent_type: Option<String>,
    pub is_parallel: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parent_task_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_ongoing: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub main_session_impact: Option<MainSessionImpact>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub team: Option<TeamMetadata>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MainSessionImpact {
    pub call_tokens: u64,
    pub result_tokens: u64,
    pub total_tokens: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TeamMetadata {
    pub team_name: String,
    pub member_name: String,
    pub member_color: String,
}

// =============================================================================
// Tool Execution
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolExecution {
    pub tool_call: ToolCall,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<ToolResult>,
    pub start_time: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub end_time: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub duration_ms: Option<f64>,
}

// =============================================================================
// Enhanced Chunk (discriminated union via chunkType tag)
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "chunkType", rename_all = "camelCase")]
pub enum EnhancedChunk {
    #[serde(rename = "user")]
    User(EnhancedUserChunk),
    #[serde(rename = "ai")]
    Ai(EnhancedAIChunk),
    #[serde(rename = "system")]
    System(EnhancedSystemChunk),
    #[serde(rename = "compact")]
    Compact(EnhancedCompactChunk),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EnhancedUserChunk {
    pub id: String,
    pub start_time: String,
    pub end_time: String,
    pub duration_ms: f64,
    pub metrics: SessionMetrics,
    pub user_message: ParsedMessage,
    pub raw_messages: Vec<ParsedMessage>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EnhancedAIChunk {
    pub id: String,
    pub start_time: String,
    pub end_time: String,
    pub duration_ms: f64,
    pub metrics: SessionMetrics,
    pub responses: Vec<ParsedMessage>,
    pub processes: Vec<Process>,
    pub sidechain_messages: Vec<ParsedMessage>,
    pub tool_executions: Vec<ToolExecution>,
    pub semantic_steps: Vec<SemanticStep>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub semantic_step_groups: Option<Vec<SemanticStepGroup>>,
    pub raw_messages: Vec<ParsedMessage>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EnhancedSystemChunk {
    pub id: String,
    pub start_time: String,
    pub end_time: String,
    pub duration_ms: f64,
    pub metrics: SessionMetrics,
    pub message: ParsedMessage,
    pub command_output: String,
    pub raw_messages: Vec<ParsedMessage>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EnhancedCompactChunk {
    pub id: String,
    pub start_time: String,
    pub end_time: String,
    pub duration_ms: f64,
    pub metrics: SessionMetrics,
    pub message: ParsedMessage,
    pub raw_messages: Vec<ParsedMessage>,
}

// =============================================================================
// Semantic Step
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SemanticStep {
    pub id: String,
    #[serde(rename = "type")]
    pub step_type: String,
    pub start_time: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub end_time: Option<String>,
    pub duration_ms: f64,
    pub content: SemanticStepContent,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tokens: Option<SemanticStepTokens>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_parallel: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub group_id: Option<String>,
    pub context: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub agent_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_message_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub effective_end_time: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub effective_duration_ms: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_gap_filled: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context_tokens: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub accumulated_context: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub token_breakdown: Option<TokenBreakdown>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SemanticStepContent {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub thinking_text: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_input: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_result_content: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_error: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_use_result: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub token_count: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub subagent_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub subagent_description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub output_text: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_model: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub interruption_text: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SemanticStepTokens {
    pub input: u64,
    pub output: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cached: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TokenBreakdown {
    pub input: u64,
    pub output: u64,
    pub cache_read: u64,
    pub cache_creation: u64,
}

// =============================================================================
// Semantic Step Group
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SemanticStepGroup {
    pub id: String,
    pub label: String,
    pub steps: Vec<SemanticStep>,
    pub is_grouped: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_message_id: Option<String>,
    pub start_time: String,
    pub end_time: String,
    pub total_duration: f64,
}

// =============================================================================
// Session Detail (complete parsed + chunked session)
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionDetail {
    pub session: super::domain::Session,
    pub messages: Vec<ParsedMessage>,
    pub chunks: Vec<EnhancedChunk>,
    pub processes: Vec<Process>,
    pub metrics: SessionMetrics,
}

File: internal/domain/chunks.go
package domain

import (
	"encoding/json"
	"fmt"
)

// Process — a subagent execution linked to a parent chunk.
type Process struct {
	ID                string             `json:"id"`
	FilePath          string             `json:"filePath"`
	Messages          []ParsedMessage    `json:"messages"`
	StartTime         string             `json:"startTime"`
	EndTime           string             `json:"endTime"`
	DurationMs        float64            `json:"durationMs"`
	Metrics           SessionMetrics     `json:"metrics"`
	Description       *string            `json:"description,omitempty"`
	SubagentType      *string            `json:"subagentType,omitempty"`
	IsParallel        bool               `json:"isParallel"`
	ParentTaskID      *string            `json:"parentTaskId,omitempty"`
	IsOngoing         *bool              `json:"isOngoing,omitempty"`
	MainSessionImpact *MainSessionImpact `json:"mainSessionImpact,omitempty"`
	Team              *TeamMetadata      `json:"team,omitempty"`
}

// MainSessionImpact — tokens a subagent cost the main session.
type MainSessionImpact struct {
	CallTokens   uint64 `json:"callTokens"`
	ResultTokens uint64 `json:"resultTokens"`
	TotalTokens  uint64 `json:"totalTokens"`
}

// TeamMetadata — agent-team enrichment for a subagent.
type TeamMetadata struct {
	TeamName    string `json:"teamName"`
	MemberName  string `json:"memberName"`
	MemberColor string `json:"memberColor"`
}

// ToolExecution — a tool_use linked to its tool_result.
type ToolExecution struct {
	ToolCall   ToolCall    `json:"toolCall"`
	Result     *ToolResult `json:"result,omitempty"`
	StartTime  string      `json:"startTime"`
	EndTime    *string     `json:"endTime,omitempty"`
	DurationMs *float64    `json:"durationMs,omitempty"`
}

// EnhancedChunk mirrors the Rust internally-tagged enum
// `#[serde(tag = "chunkType", rename_all = "camelCase")]`. Marshal emits the
// active variant's struct fields plus a "chunkType" discriminator.
type EnhancedChunk struct {
	Type    string
	User    *EnhancedUserChunk
	Ai      *EnhancedAIChunk
	System  *EnhancedSystemChunk
	Compact *EnhancedCompactChunk
	Event   *EnhancedEventChunk
}

func (c EnhancedChunk) inner() (any, error) {
	switch c.Type {
	case "user":
		return c.User, nil
	case "ai":
		return c.Ai, nil
	case "system":
		return c.System, nil
	case "compact":
		return c.Compact, nil
	case "event":
		return c.Event, nil
	default:
		return nil, fmt.Errorf("domain: unknown EnhancedChunk type %q", c.Type)
	}
}

func (c EnhancedChunk) MarshalJSON() ([]byte, error) {
	inner, err := c.inner()
	if err != nil {
		return nil, err
	}
	b, err := json.Marshal(inner)
	if err != nil {
		return nil, err
	}
	// Splice the discriminator into the object. The parity harness key-sorts
	// recursively, so the injected-key position does not affect the gate.
	var m map[string]json.RawMessage
	if err := json.Unmarshal(b, &m); err != nil {
		return nil, err
	}
	tag, _ := json.Marshal(c.Type)
	m["chunkType"] = tag
	return json.Marshal(m)
}

func (c *EnhancedChunk) UnmarshalJSON(b []byte) error {
	var head struct {
		ChunkType string `json:"chunkType"`
	}
	if err := json.Unmarshal(b, &head); err != nil {
		return err
	}
	c.Type = head.ChunkType
	switch head.ChunkType {
	case "user":
		c.User = &EnhancedUserChunk{}
		return json.Unmarshal(b, c.User)
	case "ai":
		c.Ai = &EnhancedAIChunk{}
		return json.Unmarshal(b, c.Ai)
	case "system":
		c.System = &EnhancedSystemChunk{}
		return json.Unmarshal(b, c.System)
	case "compact":
		c.Compact = &EnhancedCompactChunk{}
		return json.Unmarshal(b, c.Compact)
	case "event":
		c.Event = &EnhancedEventChunk{}
		return json.Unmarshal(b, c.Event)
	default:
		return fmt.Errorf("domain: unknown chunkType %q", head.ChunkType)
	}
}

type EnhancedUserChunk struct {
	ID          string          `json:"id"`
	StartTime   string          `json:"startTime"`
	EndTime     string          `json:"endTime"`
	DurationMs  float64         `json:"durationMs"`
	Metrics     SessionMetrics  `json:"metrics"`
	UserMessage ParsedMessage   `json:"userMessage"`
	RawMessages []ParsedMessage `json:"rawMessages"`
}

type EnhancedAIChunk struct {
	ID                string               `json:"id"`
	StartTime         string               `json:"startTime"`
	EndTime           string               `json:"endTime"`
	DurationMs        float64              `json:"durationMs"`
	Metrics           SessionMetrics       `json:"metrics"`
	Responses         []ParsedMessage      `json:"responses"`
	Processes         []Process            `json:"processes"`
	SidechainMessages []ParsedMessage      `json:"sidechainMessages"`
	ToolExecutions    []ToolExecution      `json:"toolExecutions"`
	SemanticSteps     []SemanticStep       `json:"semanticSteps"`
	SemanticStepGroups *[]SemanticStepGroup `json:"semanticStepGroups,omitempty"`
	RawMessages       []ParsedMessage      `json:"rawMessages"`
	ProgressCount     *uint32              `json:"progressCount,omitempty"`
	ProgressTexts     *[]string            `json:"progressTexts,omitempty"`
}

type EnhancedSystemChunk struct {
	ID            string          `json:"id"`
	StartTime     string          `json:"startTime"`
	EndTime       string          `json:"endTime"`
	DurationMs    float64         `json:"durationMs"`
	Metrics       SessionMetrics  `json:"metrics"`
	Message       ParsedMessage   `json:"message"`
	CommandOutput string          `json:"commandOutput"`
	RawMessages   []ParsedMessage `json:"rawMessages"`
}

type EnhancedCompactChunk struct {
	ID          string          `json:"id"`
	StartTime   string          `json:"startTime"`
	EndTime     string          `json:"endTime"`
	DurationMs  float64         `json:"durationMs"`
	Metrics     SessionMetrics  `json:"metrics"`
	Message     ParsedMessage   `json:"message"`
	RawMessages []ParsedMessage `json:"rawMessages"`
}

type EnhancedEventChunk struct {
	ID          string          `json:"id"`
	StartTime   string          `json:"startTime"`
	EndTime     string          `json:"endTime"`
	DurationMs  float64         `json:"durationMs"`
	Metrics     SessionMetrics  `json:"metrics"`
	Message     ParsedMessage   `json:"message"`
	EventData   SystemEventData `json:"eventData"`
	RawMessages []ParsedMessage `json:"rawMessages"`
}

// SemanticStep — an extracted reasoning/tool step in an AI response.
type SemanticStep struct {
	ID                  string              `json:"id"`
	StepType            string              `json:"type"`
	StartTime           string              `json:"startTime"`
	EndTime             *string             `json:"endTime,omitempty"`
	DurationMs          float64             `json:"durationMs"`
	Content             SemanticStepContent `json:"content"`
	Tokens              *SemanticStepTokens `json:"tokens,omitempty"`
	IsParallel          *bool               `json:"isParallel,omitempty"`
	GroupID             *string             `json:"groupId,omitempty"`
	Context             string              `json:"context"`
	AgentID             *string             `json:"agentId,omitempty"`
	SourceMessageID     *string             `json:"sourceMessageId,omitempty"`
	EffectiveEndTime    *string             `json:"effectiveEndTime,omitempty"`
	EffectiveDurationMs *float64            `json:"effectiveDurationMs,omitempty"`
	IsGapFilled         *bool               `json:"isGapFilled,omitempty"`
	ContextTokens       *uint64             `json:"contextTokens,omitempty"`
	AccumulatedContext  *uint64             `json:"accumulatedContext,omitempty"`
	TokenBreakdown      *TokenBreakdown     `json:"tokenBreakdown,omitempty"`
}

type SemanticStepContent struct {
	ThinkingText      *string  `json:"thinkingText,omitempty"`
	ToolName          *string  `json:"toolName,omitempty"`
	ToolInput         RawValue `json:"toolInput,omitempty"`
	ToolResultContent *string  `json:"toolResultContent,omitempty"`
	IsError           *bool    `json:"isError,omitempty"`
	ToolUseResult     RawValue `json:"toolUseResult,omitempty"`
	TokenCount        *uint64  `json:"tokenCount,omitempty"`
	SubagentID        *string  `json:"subagentId,omitempty"`
	SubagentDescription *string `json:"subagentDescription,omitempty"`
	OutputText        *string  `json:"outputText,omitempty"`
	SourceModel       *string  `json:"sourceModel,omitempty"`
	InterruptionText  *string  `json:"interruptionText,omitempty"`
}

type SemanticStepTokens struct {
	Input  uint64  `json:"input"`
	Output uint64  `json:"output"`
	Cached *uint64 `json:"cached,omitempty"`
}

type TokenBreakdown struct {
	Input         uint64 `json:"input"`
	Output        uint64 `json:"output"`
	CacheRead     uint64 `json:"cacheRead"`
	CacheCreation uint64 `json:"cacheCreation"`
}

// SemanticStepGroup — a labeled group of related steps.
type SemanticStepGroup struct {
	ID              string         `json:"id"`
	Label           string         `json:"label"`
	Steps           []SemanticStep `json:"steps"`
	IsGrouped       bool           `json:"isGrouped"`
	SourceMessageID *string        `json:"sourceMessageId,omitempty"`
	StartTime       string         `json:"startTime"`
	EndTime         string         `json:"endTime"`
	TotalDuration   float64        `json:"totalDuration"`
}

// SessionDetail — the complete parsed + chunked session (the parity-gate output).
type SessionDetail struct {
	Session  Session         `json:"session"`
	Messages []ParsedMessage `json:"messages"`
	Chunks   []EnhancedChunk `json:"chunks"`
	Processes []Process      `json:"processes"`
	Metrics  SessionMetrics  `json:"metrics"`
}
